libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -229,32 +229,53 @@ bool IRGeneratorForStatements::visit(Conditional const& _conditional)
 bool IRGeneratorForStatements::visit(Assignment const& _assignment)
 {
 	_assignment.rightHandSide().accept(*this);
-	Type const* intermediateType = type(_assignment.rightHandSide()).closestTemporaryType(
-		&type(_assignment.leftHandSide())
+
```

Comment:
I think this function would be much cleaner, if it just treated shift-assigns entirely separately - I think that way we could get around the weird ``unique_ptr<IRVariable>`` construction and wouldn't need the additional ``ostream& assign(IRVariable const&)`` construction, which pretty much goes against the design of IRVariables and undermines their iḿplicit conversion properties.

Reply:
I think most of us are still a bit puzzled about how to use the IRVariable mechanism outside of Solidity expressions. Why does `assign`  go against the design but `IRGeneratorForStatements::define` does not?

Reply:
Well, ``ostream& assign(IRVariable)`` doesn't convert types. The idea would be to instead create a temporary IRVariable, define that with the function call and assign to that temporary.
That way one needs to think about what type the function call returns, which might be a different type than what is ultimately assigned to (compared to assigning directly).

Actually in the situation here, it might make sense to skip the whole IRVariable mechanism for compound assignments, though, since all types for which that can happen will only have one stack slot, so it may be easier to just work with single yul variables for this.

Reply:
In a sense it would be better to have the util functions return temporary IRVariables to begin with - then we'd not need ``ostream& define(IRVariable)`` either - but only having ``define``, which by definition can be done only once and still needs you to give the variable an explicit type, pretty much achieves the same - that is *unless* we also add ``ostream& assign(IRVariable)``...

And rather than integrating IRVariables further into the util functions, I'd still keep the option open to remove them again, if we add native yul tuple assignments...

Reply:
Having type information associated with the yul util functions might be a good idea, especially those that already take type parameters.

Also commented on the tuple issue.

Reply:
I'm not sure this is much nicer like this.

Reply:
Ok, I think now it is nicely simplified.